### ISSUE TITLE
Schema validation incorrectly fails when using `additionalProperties: false` with `allOf`

### ISSUE THREAD

**Issue by smridge** [rswag/rswag#432]
When using `additionalProperties: false` in combination with `allOf` (and I'm guessing others such as `oneOf`) the documentation generates correctly 👍 ...However, the schema validation incorrectly fails 👎 .

For example given the following:
```ruby
resource: {
  type: :object,
  additionalProperties: false,
  allOf: [{"$ref": '#/components/schemas/foo'}],
  required: %w[bar],
  properties: {
    bar: {type: :string},
    },
```

Let's say `foobar` is referenced in `{"$ref": '#/components/schemas/foo'}` and this is expected in the response, everything is good. However, the spec itself fails when it should not:
```
     Rswag::Specs::UnexpectedResponse:
       Expected response body to match schema: The property '#/resource' contains additional properties ["foobar"] outside of the schema when none are allowed in schema
```

I'm not too familiar with this repo but it looks like [this](https://github.com/ruby-json-schema/json-schema/issues/266) (thanks @davidgm0 for finding!) might be the culprit this repo uses to validate?

Yes, we can avoid this failure by removing `additionalProperties: false`. However, this is not ideal when the goal is to enforce documentation accuracy as much as possible.

**Comment by smridge** [comment]
I realized this was answered in the issue I referenced above , specifically in [subschema-independence](https://json-schema.org/understanding-json-schema/reference/combining.html#subschema-independence) (thanks @bopm for the reference!).
> Unfortunately, now the schema will reject everything. This is because additionalProperties knows nothing about the properties declared in the subschemas inside of the allOf array.
To many, this is one of the biggest surprises of the combining operations in JSON schema: it does not behave like inheritance in an object-oriented language. There are some proposals to address this in the next version of the JSON schema specification.

With that said, I can close this issue.
